<a href="https://colab.research.google.com/github/KomisD/Drug-Classification/blob/main/Project3_NNs.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Introduction

- **The Dataset:** https://www.kaggle.com/datasets/prathamtripathi/drug-classification/data
- **Objectives of the analysis:** I tryied several methodes to find the best "doctor" model, so it can predict based on the features what type of drug a patient needs.


# Importing the libraries

In [ ]:
pip install category_encoders

In [ ]:
import numpy as np
import matplotlib.pyplot as plt
from sklearn.preprocessing import LabelEncoder
from sklearn.model_selection import train_test_split, KFold, cross_val_score
import pandas as pd
import category_encoders as ce

## Importing the dataset and encoding categorical values


In [ ]:
dataset = pd.read_csv('drug200.csv')

# Applying the binary encoder for SEX
binary_encoder = ce.BinaryEncoder(cols=['Sex'])
dataset = binary_encoder.fit_transform(dataset)

# Applying the OneHot encoder for blood pressure
OneHot = ce.OneHotEncoder(cols=['BP'])
dataset = OneHot.fit_transform(dataset)

#Applaying binary encoder for Cholesterol
binary_encoder = ce.BinaryEncoder(cols=['Cholesterol'])
dataset = binary_encoder.fit_transform(dataset)

#spliting variables into depended and target value
X = dataset.iloc[:, :-1].values
y = dataset.iloc[:, -1].values

# Applying label encoding to the target variable 'Drug'
label_encoder = LabelEncoder()
y = label_encoder.fit_transform(y)

## Splitting the dataset into the Training set and Test set

In [ ]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.4)

In [ ]:
print(X_train)

[[50.     1.     0.    ...  1.     0.    15.79 ]
 [28.     1.     0.    ...  0.     1.    27.064]
 [32.     0.     1.    ...  1.     0.    10.292]
 ...
 [65.     0.     1.    ...  1.     0.    31.876]
 [54.     1.     0.    ...  0.     1.    24.658]
 [38.     0.     1.    ...  1.     0.    11.326]]


In [ ]:
print(y_train)

[0 0 1 3 1 4 4 2 0 4 3 1 0 0 4 2 0 0 1 4 4 0 0 0 3 0 0 3 0 0 4 0 0 4 4 4 4
 2 1 0 0 0 4 1 0 4 4 4 0 0 4 0 0 3 4 0 4 3 0 0 4 0 0 0 4 4 3 0 3 4 0 4 3 0
 0 0 0 0 0 0 2 0 1 2 0 0 0 2 3 0 0 1 0 0 0 0 4 1 4 2 0 4 0 4 0 0 4 4 0 0 2
 4 1 4 0 0 3 0 0 1]


In [ ]:
print(X_test)

[[19.     0.     1.     1.     0.     0.     0.     1.    13.313]
 [57.     0.     1.     0.     0.     1.     0.     1.    14.216]
 [36.     0.     1.     0.     0.     1.     0.     1.    16.753]
 [69.     1.     0.     0.     1.     0.     0.     1.    15.478]
 [53.     0.     1.     1.     0.     0.     1.     0.    12.495]
 [28.     0.     1.     1.     0.     0.     1.     0.    18.809]
 [32.     0.     1.     1.     0.     0.     1.     0.    25.974]
 [34.     1.     0.     0.     0.     1.     0.     1.    22.456]
 [20.     0.     1.     0.     1.     0.     1.     0.    11.686]
 [28.     0.     1.     0.     0.     1.     0.     1.    12.879]
 [51.     0.     1.     0.     0.     1.     0.     1.    13.597]
 [58.     0.     1.     0.     1.     0.     0.     1.    26.645]
 [70.     1.     0.     1.     0.     0.     0.     1.    13.967]
 [37.     1.     0.     0.     1.     0.     1.     0.     8.968]
 [31.     1.     0.     1.     0.     0.     1.     0.    11.871]
 [60.     

In [ ]:
print(y_test)

[1 4 0 0 2 0 0 0 4 4 4 0 2 4 1 4 2 4 1 0 4 0 4 2 0 0 4 2 1 0 4 0 2 4 0 0 0
 0 1 1 0 0 1 1 4 0 3 0 0 0 0 2 3 3 0 4 3 1 4 0 4 0 0 4 4 0 4 4 0 4 2 0 1 1
 1 0 0 3 4 4]


## Feature Scaling

In [ ]:
from sklearn.preprocessing import StandardScaler
sc = StandardScaler()
X_train = sc.fit_transform(X_train)
X_test = sc.transform(X_test)

In [ ]:
print(X_train)

[[ 0.35586833  1.03390789 -1.03390789 ...  0.91986621 -0.91986621
  -0.09993652]
 [-0.95627187  1.03390789 -1.03390789 ... -1.08711461  1.08711461
   1.41054975]
 [-0.71770093 -0.96720415  0.96720415 ...  0.91986621 -0.91986621
  -0.83655648]
 ...
 [ 1.25050937 -0.96720415  0.96720415 ...  0.91986621 -0.91986621
   2.0552597 ]
 [ 0.59443927  1.03390789 -1.03390789 ... -1.08711461  1.08711461
   1.08819478]
 [-0.35984451 -0.96720415  0.96720415 ...  0.91986621 -0.91986621
  -0.69802155]]


In [ ]:
print(X_test)

[[-1.4930565  -0.96720415  0.96720415  1.26834366 -0.70710678 -0.62876771
  -1.08711461  1.08711461 -0.43180404]
 [ 0.77336748 -0.96720415  0.96720415 -0.78842985 -0.70710678  1.59041245
  -1.08711461  1.08711461 -0.31082044]
 [-0.47912998 -0.96720415  0.96720415 -0.78842985 -0.70710678  1.59041245
  -1.08711461  1.08711461  0.02908586]
 [ 1.48908032  1.03390789 -1.03390789 -0.78842985  1.41421356 -0.62876771
  -1.08711461  1.08711461 -0.14173816]
 [ 0.53479654 -0.96720415  0.96720415  1.26834366 -0.70710678 -0.62876771
   0.91986621 -0.91986621 -0.54139937]
 [-0.95627187 -0.96720415  0.96720415  1.26834366 -0.70710678 -0.62876771
   0.91986621 -0.91986621  0.30454796]
 [-0.71770093 -0.96720415  0.96720415  1.26834366 -0.70710678 -0.62876771
   0.91986621 -0.91986621  1.26451196]
 [-0.59841546  1.03390789 -1.03390789 -0.78842985 -0.70710678  1.59041245
  -1.08711461  1.08711461  0.79317165]
 [-1.43341377 -0.96720415  0.96720415 -0.78842985  1.41421356 -0.62876771
   0.91986621 -0.91986

# RBFN

## Simple Implimentation


Here from the trainable parameters I set a random beta and the Prototypes random initialized to just have a working model

In [ ]:
from keras.layers import Layer
from keras import backend as K
from keras.models import Sequential
from keras.layers import Dense
from keras.losses import sparse_categorical_crossentropy

class RBFLayer(Layer):
    def __init__(self, units, gamma, **kwargs):
        super(RBFLayer, self).__init__(**kwargs)
        self.units = units
        self.gamma = K.cast_to_floatx(gamma)

    def build(self, input_shape):
        self.mu = self.add_weight(name='mu',
                                  shape=(int(input_shape[1]), self.units),
                                  initializer='uniform',
                                  trainable=True)
        super(RBFLayer, self).build(input_shape)

    def call(self, inputs):
        diff = K.expand_dims(inputs) - self.mu
        l2 = K.sum(K.pow(diff, 2), axis=1)
        res = K.exp(-1 * self.gamma * l2)
        return res

In [ ]:
model = Sequential()
model.add(RBFLayer(20, 0.5, input_shape=(9,)))  # 10 RBF units, adjust as needed
model.add(Dense(5, activation='softmax', name='output'))  # 5 output neurons for 5 classes

In [156]:
# Compile the model
model.compile(optimizer='rmsprop', loss='sparse_categorical_crossentropy', metrics=['accuracy'])

# Fit the model
model.fit(X_train, y_train, batch_size=8, epochs=400)

Epoch 1/400
15/15 [==============================] - 1s 5ms/step - loss: 0.3218 - accuracy: 0.9000
Epoch 2/400
15/15 [==============================] - 0s 3ms/step - loss: 0.3204 - accuracy: 0.9000
Epoch 3/400
15/15 [==============================] - 0s 4ms/step - loss: 0.3195 - accuracy: 0.9000
Epoch 4/400
15/15 [==============================] - 0s 6ms/step - loss: 0.3187 - accuracy: 0.9000
Epoch 5/400
15/15 [==============================] - 0s 5ms/step - loss: 0.3181 - accuracy: 0.9000
Epoch 6/400
15/15 [==============================] - 0s 9ms/step - loss: 0.3174 - accuracy: 0.9000
Epoch 7/400
15/15 [==============================] - 0s 7ms/step - loss: 0.3165 - accuracy: 0.9000
Epoch 8/400
15/15 [==============================] - 0s 11ms/step - loss: 0.3159 - accuracy: 0.9000
Epoch 9/400
15/15 [==============================] - 0s 5ms/step - loss: 0.3152 - accuracy: 0.9000
Epoch 10/400
15/15 [==============================] - 0s 5ms/step - loss: 0.3145 - accuracy: 0.9000
Epoch 11

In [157]:
# Evaluate the model on the test set
loss, accuracy = model.evaluate(X_test, y_test, verbose=0)

# Print the accuracy
print(f'Accuracy: {accuracy * 100:.2f}%')

Accuracy: 80.00%


## Training it with different parameters

After this implimentation I will try to find the Prototypes using KMeans clustering and I will also set the correct beta value based on the equation from theory

In [ ]:
from sklearn.cluster import KMeans

# Number of RBF neurons
num_rbf_neurons = 10
kmeans = KMeans(n_clusters=num_rbf_neurons, random_state=0).fit(X_train)
prototypes = kmeans.cluster_centers_

/usr/local/lib/python3.10/dist-packages/sklearn/cluster/_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning
  warnings.warn(


In [ ]:
print(prototypes)

[[-0.01902887  1.03390789 -1.03390789  1.26834366 -0.70710678 -0.62876771
   0.91986621 -0.91986621 -0.01094534]
 [-0.24055904 -0.96720415  0.96720415  0.92554808 -0.35355339 -0.62876771
  -1.08711461  1.08711461  1.27016144]
 [ 0.09493136  1.03390789 -1.03390789 -0.78842985 -0.70710678  1.59041245
  -1.08711461  1.08711461  0.33583221]
 [ 0.22031665  1.03390789 -1.03390789 -0.78842985  1.41421356 -0.62876771
   0.91986621 -0.91986621  0.04109531]
 [-0.05765465 -0.03335187  0.03335187 -0.78842985 -0.70710678  1.59041245
   0.91986621 -0.91986621 -0.52081117]
 [-0.50895135 -0.96720415  0.96720415  1.26834366 -0.70710678 -0.62876771
   0.91986621 -0.91986621  0.25752112]
 [ 0.07256533  0.28349087 -0.28349087 -0.78842985  1.41421356 -0.62876771
  -1.08711461  1.08711461 -0.49708561]
 [ 0.11729738 -0.96720415  0.96720415 -0.78842985  1.41421356 -0.62876771
   0.91986621 -0.91986621  0.07043684]
 [ 0.36332367  1.03390789 -1.03390789  1.26834366 -0.70710678 -0.62876771
  -1.08711461  1.08711

In [ ]:
from sklearn.cluster import KMeans
from keras.layers import Layer
from keras import backend as K
import numpy as np
from keras.initializers import Constant

class RBFLayer2(Layer):
    def __init__(self, units, gamma, initializer, **kwargs):
        super(RBFLayer2, self).__init__(**kwargs)
        self.units = units
        self.gamma = K.cast_to_floatx(gamma)
        self.initializer = initializer

    def build(self, input_shape):
        self.mu = self.add_weight(name='mu',
                                  shape=(int(input_shape[1]), self.units),
                                  initializer=self.initializer,
                                  trainable=True)
        super(RBFLayer2, self).build(input_shape)

    def call(self, inputs):
        diff = K.expand_dims(inputs) - self.mu
        l2 = K.sum(K.pow(diff, 2), axis=1)
        res = K.exp(-1 * self.gamma * l2)
        return res

    def compute_output_shape(self, input_shape):
        return (input_shape[0], self.units)


In [153]:
from keras.layers import Layer
from keras.models import Sequential
from keras.layers import Dense
from keras.initializers import Constant
import numpy as np
from sklearn.cluster import KMeans

#calculating the beta value
data_variance = np.var(X_train, axis=0).mean()
gamma = 1 / (2 * data_variance)

model = Sequential()
model.add(RBFLayer2(units=num_rbf_neurons, gamma=gamma, initializer=Constant(prototypes.T), input_shape=(X_train.shape[1],)))
model.add(Dense(5, activation='softmax', name='output'))

In [154]:
# Compile the model
model.compile(optimizer='rmsprop', loss='sparse_categorical_crossentropy', metrics=['accuracy'])

# Fit the model
model.fit(X_train, y_train, batch_size=8, epochs=400)

Epoch 1/400
15/15 [==============================] - 0s 2ms/step - loss: 1.6561 - accuracy: 0.1750
Epoch 2/400
15/15 [==============================] - 0s 3ms/step - loss: 1.6350 - accuracy: 0.2167
Epoch 3/400
15/15 [==============================] - 0s 2ms/step - loss: 1.6171 - accuracy: 0.2500
Epoch 4/400
15/15 [==============================] - 0s 2ms/step - loss: 1.6000 - accuracy: 0.2417
Epoch 5/400
15/15 [==============================] - 0s 2ms/step - loss: 1.5833 - accuracy: 0.2750
Epoch 6/400
15/15 [==============================] - 0s 2ms/step - loss: 1.5668 - accuracy: 0.3250
Epoch 7/400
15/15 [==============================] - 0s 2ms/step - loss: 1.5508 - accuracy: 0.3667
Epoch 8/400
15/15 [==============================] - 0s 3ms/step - loss: 1.5351 - accuracy: 0.3750
Epoch 9/400
15/15 [==============================] - 0s 2ms/step - loss: 1.5193 - accuracy: 0.3917
Epoch 10/400
15/15 [==============================] - 0s 2ms/step - loss: 1.5041 - accuracy: 0.3917
Epoch 11/

In [155]:
# Evaluate the model on the test set
loss, accuracy = model.evaluate(X_test, y_test, verbose=0)

# Print the accuracy
print(f'Accuracy: {accuracy * 100:.2f}%')

Accuracy: 70.00%


## Genering training and experimentation

Here I will try compine those 2 aproches from above, trying different type of neurons, gamma, batch size, epochs and a lot of experimentation everything will be based to RBFLayer and RBFLayer2

In [ ]:
# different number of neurons to the hidden layer
model = Sequential()
model.add(RBFLayer(30, 0.5, input_shape=(9,)))
model.add(Dense(5, activation='softmax', name='output'))
model.compile(optimizer='rmsprop', loss='sparse_categorical_crossentropy', metrics=['accuracy'])
model.fit(X_train, y_train, batch_size=8, epochs=400)

Epoch 1/400
15/15 [==============================] - 1s 2ms/step - loss: 1.5981 - accuracy: 0.3667
Epoch 2/400
15/15 [==============================] - 0s 2ms/step - loss: 1.5841 - accuracy: 0.4167
Epoch 3/400
15/15 [==============================] - 0s 3ms/step - loss: 1.5728 - accuracy: 0.5000
Epoch 4/400
15/15 [==============================] - 0s 6ms/step - loss: 1.5619 - accuracy: 0.5000
Epoch 5/400
15/15 [==============================] - 0s 5ms/step - loss: 1.5512 - accuracy: 0.4917
Epoch 6/400
15/15 [==============================] - 0s 7ms/step - loss: 1.5408 - accuracy: 0.4917
Epoch 7/400
15/15 [==============================] - 0s 5ms/step - loss: 1.5305 - accuracy: 0.4917
Epoch 8/400
15/15 [==============================] - 0s 4ms/step - loss: 1.5208 - accuracy: 0.4917
Epoch 9/400
15/15 [==============================] - 0s 4ms/step - loss: 1.5112 - accuracy: 0.4917
Epoch 10/400
15/15 [==============================] - 0s 4ms/step - loss: 1.5016 - accuracy: 0.4917
Epoch 11/

In [ ]:
# Evaluate the model on the test set
loss, accuracy = model.evaluate(X_test, y_test, verbose=0)

# Print the accuracy
print(f'Accuracy: {accuracy * 100:.2f}%')

Accuracy: 60.00%


In [ ]:
# different number of neurons to the hidden layer to 50
model = Sequential()
model.add(RBFLayer(50, 0.5, input_shape=(9,)))
model.add(Dense(5, activation='softmax', name='output'))
model.compile(optimizer='rmsprop', loss='sparse_categorical_crossentropy', metrics=['accuracy'])
model.fit(X_train, y_train, batch_size=8, epochs=400)

Epoch 1/400
15/15 [==============================] - 1s 2ms/step - loss: 1.5961 - accuracy: 0.4250
Epoch 2/400
15/15 [==============================] - 0s 2ms/step - loss: 1.5787 - accuracy: 0.4917
Epoch 3/400
15/15 [==============================] - 0s 2ms/step - loss: 1.5646 - accuracy: 0.4917
Epoch 4/400
15/15 [==============================] - 0s 3ms/step - loss: 1.5517 - accuracy: 0.4917
Epoch 5/400
15/15 [==============================] - 0s 2ms/step - loss: 1.5393 - accuracy: 0.4917
Epoch 6/400
15/15 [==============================] - 0s 2ms/step - loss: 1.5268 - accuracy: 0.4917
Epoch 7/400
15/15 [==============================] - 0s 2ms/step - loss: 1.5143 - accuracy: 0.4917
Epoch 8/400
15/15 [==============================] - 0s 3ms/step - loss: 1.5026 - accuracy: 0.4917
Epoch 9/400
15/15 [==============================] - 0s 2ms/step - loss: 1.4916 - accuracy: 0.4917
Epoch 10/400
15/15 [==============================] - 0s 2ms/step - loss: 1.4807 - accuracy: 0.4917
Epoch 11/

In [ ]:
# Evaluate the model on the test set
loss, accuracy = model.evaluate(X_test, y_test, verbose=0)

# Print the accuracy
print(f'Accuracy: {accuracy * 100:.2f}%')

Accuracy: 76.25%


In [ ]:
# here I calculate KMeans clustering with more neurons
from sklearn.cluster import KMeans

# Number of RBF neurons
num_rbf_neurons = 30
kmeans = KMeans(n_clusters=num_rbf_neurons, random_state=0).fit(X_train)
prototypes = kmeans.cluster_centers_

model = Sequential()
model.add(RBFLayer2(units=num_rbf_neurons, gamma=gamma, initializer=Constant(prototypes.T), input_shape=(X_train.shape[1],)))
model.add(Dense(5, activation='softmax', name='output'))
model.compile(optimizer='rmsprop', loss='sparse_categorical_crossentropy', metrics=['accuracy'])
model.fit(X_train, y_train, batch_size=8, epochs=400)

Epoch 1/400


/usr/local/lib/python3.10/dist-packages/sklearn/cluster/_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning
  warnings.warn(


15/15 [==============================] - 1s 3ms/step - loss: 1.6380 - accuracy: 0.2167
Epoch 2/400
15/15 [==============================] - 0s 4ms/step - loss: 1.6165 - accuracy: 0.2417
Epoch 3/400
15/15 [==============================] - 0s 4ms/step - loss: 1.5986 - accuracy: 0.2583
Epoch 4/400
15/15 [==============================] - 0s 4ms/step - loss: 1.5815 - accuracy: 0.2750
Epoch 5/400
15/15 [==============================] - 0s 4ms/step - loss: 1.5645 - accuracy: 0.2833
Epoch 6/400
15/15 [==============================] - 0s 3ms/step - loss: 1.5477 - accuracy: 0.3250
Epoch 7/400
15/15 [==============================] - 0s 3ms/step - loss: 1.5318 - accuracy: 0.3500
Epoch 8/400
15/15 [==============================] - 0s 4ms/step - loss: 1.5165 - accuracy: 0.4000
Epoch 9/400
15/15 [==============================] - 0s 8ms/step - loss: 1.5017 - accuracy: 0.4167
Epoch 10/400
15/15 [==============================] - 0s 22ms/step - loss: 1.4867 - accuracy: 0.4167
Epoch 11/400
15/15 [

In [ ]:
# Evaluate the model on the test set
loss, accuracy = model.evaluate(X_test, y_test, verbose=0)

# Print the accuracy
print(f'Accuracy: {accuracy * 100:.2f}%')

Accuracy: 68.75%


In [ ]:
#Here I will try different batch size
model = Sequential()
model.add(RBFLayer(30, 0.5, input_shape=(9,)))
model.add(Dense(5, activation='softmax', name='output'))
model.compile(optimizer='rmsprop', loss='sparse_categorical_crossentropy', metrics=['accuracy'])
model.fit(X_train, y_train, batch_size=16, epochs=400)

Epoch 1/400
8/8 [==============================] - 0s 3ms/step - loss: 1.6117 - accuracy: 0.1917
Epoch 2/400
8/8 [==============================] - 0s 3ms/step - loss: 1.6017 - accuracy: 0.3417
Epoch 3/400
8/8 [==============================] - 0s 3ms/step - loss: 1.5935 - accuracy: 0.4833
Epoch 4/400
8/8 [==============================] - 0s 2ms/step - loss: 1.5863 - accuracy: 0.4917
Epoch 5/400
8/8 [==============================] - 0s 4ms/step - loss: 1.5791 - accuracy: 0.4917
Epoch 6/400
8/8 [==============================] - 0s 4ms/step - loss: 1.5722 - accuracy: 0.4917
Epoch 7/400
8/8 [==============================] - 0s 3ms/step - loss: 1.5655 - accuracy: 0.4917
Epoch 8/400
8/8 [==============================] - 0s 3ms/step - loss: 1.5590 - accuracy: 0.4917
Epoch 9/400
8/8 [==============================] - 0s 3ms/step - loss: 1.5526 - accuracy: 0.4917
Epoch 10/400
8/8 [==============================] - 0s 3ms/step - loss: 1.5463 - accuracy: 0.4917
Epoch 11/400
8/8 [===========

In [ ]:
# Evaluate the model on the test set
loss, accuracy = model.evaluate(X_test, y_test, verbose=0)

# Print the accuracy
print(f'Accuracy: {accuracy * 100:.2f}%')

Accuracy: 58.75%


In [ ]:
#The same batch size for the model with the KMeans Prototypes
from sklearn.cluster import KMeans

# Number of RBF neurons
num_rbf_neurons = 30
kmeans = KMeans(n_clusters=num_rbf_neurons, random_state=0).fit(X_train)
prototypes = kmeans.cluster_centers_

model = Sequential()
model.add(RBFLayer2(units=num_rbf_neurons, gamma=gamma, initializer=Constant(prototypes.T), input_shape=(X_train.shape[1],)))
model.add(Dense(5, activation='softmax', name='output'))
model.compile(optimizer='rmsprop', loss='sparse_categorical_crossentropy', metrics=['accuracy'])
model.fit(X_train, y_train, batch_size=16, epochs=400)

/usr/local/lib/python3.10/dist-packages/sklearn/cluster/_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning
  warnings.warn(


Epoch 1/400
8/8 [==============================] - 0s 3ms/step - loss: 1.6267 - accuracy: 0.2250
Epoch 2/400
8/8 [==============================] - 0s 2ms/step - loss: 1.6106 - accuracy: 0.2333
Epoch 3/400
8/8 [==============================] - 0s 4ms/step - loss: 1.5980 - accuracy: 0.2500
Epoch 4/400
8/8 [==============================] - 0s 2ms/step - loss: 1.5868 - accuracy: 0.2667
Epoch 5/400
8/8 [==============================] - 0s 3ms/step - loss: 1.5759 - accuracy: 0.2833
Epoch 6/400
8/8 [==============================] - 0s 2ms/step - loss: 1.5653 - accuracy: 0.3000
Epoch 7/400
8/8 [==============================] - 0s 2ms/step - loss: 1.5548 - accuracy: 0.3333
Epoch 8/400
8/8 [==============================] - 0s 3ms/step - loss: 1.5447 - accuracy: 0.3500
Epoch 9/400
8/8 [==============================] - 0s 2ms/step - loss: 1.5350 - accuracy: 0.3583
Epoch 10/400
8/8 [==============================] - 0s 3ms/step - loss: 1.5249 - accuracy: 0.3667
Epoch 11/400
8/8 [===========

In [ ]:
# Evaluate the model on the test set
loss, accuracy = model.evaluate(X_test, y_test, verbose=0)

# Print the accuracy
print(f'Accuracy: {accuracy * 100:.2f}%')

Accuracy: 65.00%


In [ ]:
# different number of neurons to the hidden layer to 50
model = Sequential()
model.add(RBFLayer(50, gamma, input_shape=(9,)))
model.add(Dense(5, activation='softmax', name='output'))
model.compile(optimizer='rmsprop', loss='sparse_categorical_crossentropy', metrics=['accuracy'])
model.fit(X_train, y_train, batch_size=8, epochs=700)

Epoch 1/700
15/15 [==============================] - 2s 5ms/step - loss: 1.5863 - accuracy: 0.4917
Epoch 2/700
15/15 [==============================] - 0s 7ms/step - loss: 1.5709 - accuracy: 0.4917
Epoch 3/700
15/15 [==============================] - 0s 6ms/step - loss: 1.5582 - accuracy: 0.4917
Epoch 4/700
15/15 [==============================] - 0s 3ms/step - loss: 1.5462 - accuracy: 0.4917
Epoch 5/700
15/15 [==============================] - 0s 5ms/step - loss: 1.5340 - accuracy: 0.4917
Epoch 6/700
15/15 [==============================] - 0s 8ms/step - loss: 1.5222 - accuracy: 0.4917
Epoch 7/700
15/15 [==============================] - 0s 8ms/step - loss: 1.5108 - accuracy: 0.4917
Epoch 8/700
15/15 [==============================] - 0s 9ms/step - loss: 1.4998 - accuracy: 0.4917
Epoch 9/700
15/15 [==============================] - 0s 5ms/step - loss: 1.4891 - accuracy: 0.4917
Epoch 10/700
15/15 [==============================] - 0s 6ms/step - loss: 1.4786 - accuracy: 0.4917
Epoch 11/

In [ ]:
# Evaluate the model on the test set
loss, accuracy = model.evaluate(X_test, y_test, verbose=0)

# Print the accuracy
print(f'Accuracy: {accuracy * 100:.2f}%')

Accuracy: 88.75%


# Applying the other Models (KNN, NCC)

## Training the KNN model on the Training set

In [ ]:
from sklearn.neighbors import KNeighborsClassifier
classifierKNN = KNeighborsClassifier(n_neighbors = 5, metric = 'minkowski', p = 2)
classifierKNN.fit(X_train, y_train)

KNeighborsClassifier()

##Training the Nearest Class Centroid model

In [ ]:
from sklearn.neighbors import NearestCentroid
classifierNC = NearestCentroid(metric='euclidean')
classifierNC.fit(X_train, y_train)

NearestCentroid()

## Predicting the Test set results

In [ ]:
y_pred_KNN = classifierKNN.predict(X_test)
y_pred_NC = classifierNC.predict(X_test)

## Making the Confusion Matrix

In [ ]:
from sklearn.metrics import accuracy_score
print("The accurasy score of KNN Model is: ", accuracy_score(y_test, y_pred_KNN) )
print("The accurasy score of NCC Model is: ", accuracy_score(y_test, y_pred_NC) )

The accurasy score of KNN Model is:  0.575
The accurasy score of NCC Model is:  0.725


# Methodology

### Preprocessing Steps

1. **Encoding Categorical Values**:
   - **Binary Encoding**: Applied to the 'Sex' and 'Cholesterol' columns.
   - **One-Hot Encoding**: Used for the 'BP' (Blood Pressure) column.

2. **Feature Scaling**:
   - **StandardScaler**: Standardizes features by removing the mean and scaling to unit variance.

### Parameter Selection and Model Tuning Methods

1. **Radial Basis Function (RBFN)**:
   - Neuron Count: Experimented with different numbers of neurons in the hidden layer (e.g., 20, 30, 50). This parameter affects the capacity of the model to capture complex patterns.
   - Gamma Value: Calculated based on data variance to determine the spread of the RBF or some constant values ranging from 0.5 to 2.
   - Prototype Initialization: In some models, prototypes were initialized using KMeans clustering, which provides a more informed starting point for these centers and in others randomly select prototypes from the training data.

2. **K-Nearest Neighbors (KNN)**:
   - **Number of Neighbors (n_neighbors)**: 5
   - **Distance Metric**: Minkowski with \( p = 2 \) (Euclidean distance)

3. **Nearest Class Centroid (NCC)**:
   - **Metric**: Euclidean distance

# Algorithm Explanation

### Radial Basis Function (RBFN)

- **Overview**: RBFNs are a type of neural network that uses radial basis functions as activation functions. They are particularly adept at interpolation in multidimensional space.
- **Training Process**: Involves adjusting both the centers of the radial basis functions (prototypes) and the weights of the output layer.
Prototype Initialization: Prototypes are either randomly initialized or set using k-means clustering.

### Nearest Neighbor (KNN)

- **Basic Concept**: Classifies a data point based on how its neighbors are classified.
- **Distance Metrics**: Often uses Euclidean distance but can use other metrics.
- **Choosing 'K'**: The number of neighbors (K) is a crucial parameter, with a larger K smoothing the decision boundaries. Also involves selecting gamma Values

### Nearest Class Centroid (NCC)

- **Principle**: Assigns a class based on the closest class centroid.
- **Centroid Calculation**: The centroid of a class is the average of all points in that class.
- **Suitability**: Effective when classes are compact and well-separated.


# Results

- RBFN Performance:
Models with different neuron counts in the hidden layer showed varying accuracy levels, with the best performance observed at `88.75`% accuracy for 50 neurons.
The implementation using KMeans clustering for prototype initialization didn't demonstrate an improvement to the randomly initialized prototypes that achives this highest accuracy.
- **Comparative analysis with Nearest Neighbor and Nearest Class Centroid classifiers:** For NC accuracy is `0.7625` and for KNN `0.7`

# Discussion
- RBFN Insights: The experiments with the Radial Basis Function Network (RBFN) showed that the model's performance is significantly influenced by the number of neurons in the hidden layer. The best performance was observed at 88.75% accuracy with 50 neurons. This indicates that a higher number of neurons in the RBF layer allows the network to capture more complex patterns and relationships in the data, leading to improved classification accuracy.

- Prototype Initialization: The implementation using KMeans clustering for prototype initialization did not demonstrate an improvement over the models with randomly initialized prototypes. It might indicate that the dataset characteristics or the model's configuration are such that random initialization already provides a sufficiently good starting point for optimization.

- Model Comparisons: Compared to the K-Nearest Neighbors (KNN) and Nearest Class Centroid (NCC) models, the RBFN with 50 neurons showed superior performance. While KNN and NCC are simpler and more straightforward to implement, they were outperformed by the more complex and adaptable RBFN model, especially when the network had a higher neuron count.
